admin/server.go

```diff
@@ -0,0 +1,55 @@
+package admin
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"github.com/reddit/baseplate.go/log"
+)
+
+const (
+	DefaultAdminAddr = ":6060"
+)
+
+// ServerArgs contain optional configuration for the admin server.
+type ServerArgs struct {
+	// AdminAddr is a custom address for the admin server. Defaults to port 6060 on localhost if not set.
+	AdminAddr string
+	// HealthCheckFn is the HTTP handler for health check.
+	HealthCheckFn func(w http.ResponseWriter, r *http.Request)
+}
+
+// NewServer returns a new admin server for internal functionality.
+func NewServer(args *ServerArgs) *Server {
+	adminAddr := args.AdminAddr
+	if args.AdminAddr == "" {
+		adminAddr = DefaultAdminAddr
+	}
+
+	return &Server{
+		adminAddr:     adminAddr,
+		healthCheckFn: args.HealthCheckFn,
+	}
+}
+
+type Server struct {
+	adminAddr     string
+	healthCheckFn func(w http.ResponseWriter, r *http.Request)
+}
+
+// Serve starts an HTTP server for internal functions:
+//    metrics       - serve /metrics for prometheus
+//    health check  - serve /health for health checking
+//    pprof         - https://blog.golang.org/pprof
```

Comment:
so this does not happen with this function. the service still needs to do the blank import in their main package, and we can not do that for them here (because blank imports shall only happen in main package), which means this is techincally... a lie.

I'm not sure what's a good way to handle this, though.

Reply:
Instead of doing a blank import, we could explicitly setup pprof by explicitly setting up the http handlers in `Serve()` rather than depend on the main `init()`.

https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/net/http/pprof/pprof.go;l=80-86

Reply:
I considered that, but that means if in the future a new pprof endpoint is added to go, people won't get them automatically by updating go, and we have to add them here.

and I'm not sure what would happen if they also blank import it in their main (e.g. the endpoints are double registered)

Reply:
Yea, it's a bit more work for us, but I don't think it's that bad. That `init()` section hasn't changed since 2017, so not much churn. 

Yea, I thought about the double registering. I guess we could do some tests.

Reply:
I have a slight preference for copying the pprof `init` func into this PR.  But I don't know what double registering would do. I will test it out.

Otherwise, we can make sure the blank import is happening in baseplate cookiecutter.

Reply:
I would prefer to change the documentation to something like "this also provides pprof if you also have blank import in your main".

alternatively we can use a non-default mux and copy pprof's `init` here, which avoids the double registration problem. but we still have the "if upstream adds a new useful pprof endpoint we have to add the support manually here" problem.

Reply:
double registration results in this panic `panic: http: multiple registrations for /debug/pprof/`.

We could use the non-default mux and copy pprof's init, but I agree @fishy I prefer to add documentation in the comment that says to "add pprof if you also have blank import in your main".  Will do that.

Reply:
Not serving /debug/pprof unless you have a magic import in your main is unacceptable for this kind of "fire and forget" API.

We have a few options:
* Document it as a side-effect of importing the package
* Serve them on a different path
* Use our own servemux

I would go with the last one: make our own top-level AdminMux and init()-register the /debug/ handlers.  The init changes VERY infrequently because only special-handling profiles need to be added to it, Index dynamically lists and dispatches most new profiles that would be added.

Reply:
Sounds good. I will add our own AdminMux.

Reply:
Yes, having a separate AdminMux is probably the best option. It makes sure we explicitly separate traffic to the admin port.